stan-dev/math
# Feature/issue 353 cov exp quad grad

#### Submisison Checklist


#### Summary:

Specialize gradients for cov_exp_quad:

- `std::vector<double> x, var sigma, var l`
- `std::vector<double> x, double sigma, var l` 

#### Intended Effect:

Speed up gradients by about 40% to 50%

#### How to Verify:

Unfortunately, the only way to verify is to exclude `stan/math/rev/mat/fun/cov_exp_quad.hpp` from the includes and run something like below, record the time, and then include `stan/math/rev/mat/fun/cov_exp_quad.hpp`  and look at the diff. Came out to somewhere between 40% and 50% depending on the size of the matrix:

```
#include <stan/math/rev/mat.hpp>
#include <gtest/gtest.h>
#include <test/unit/math/rev/mat/fun/util.hpp>
#include <sys/time.h>
#include <ctime>

// returns milliseconds
unsigned int get_time(){
  struct timeval tv;

  gettimeofday(&tv,NULL);

  unsigned int ret = tv.tv_usec;
  ret /= 1000;

  ret += (tv.tv_sec * 1000);

  return ret;
}

TEST(AgradRevMatrix, cov_exp_quad_timing) {
  using stan::math::matrix_v;
  using stan::math::var;
  using stan::math::cov_exp_quad;
  using std::vector;

  unsigned int start, end;
  int n = 1e2;
  double sum_grad(0.0);
  int size = 500;
  Eigen::VectorXd dv = Eigen::VectorXd::Random(size);
  vector<double> vv;
  for (int i = 0; i < size; ++i)
    vv.push_back(dv(i));

  start = get_time();
  for (int k = 0; k < n; ++k) {
    var sig = 1;
    var len = 2;
    matrix_v cov_mat = cov_exp_quad(vv, sig, len);

    AVEC ab;
    VEC grad;
    ab.push_back(sig);
    ab.push_back(len);
    
    cov_mat(0,1).grad(ab, grad);
    for (size_t j = 0; j < static_cast<size_t>(grad.size()); ++j)
      sum_grad += grad[j];
    stan::math::recover_memory();
  }
  end = get_time();
  std::cout << sum_grad << std::endl;
  std::cout << end - start << std::endl;
}
```

#### Side Effects:

None.

#### Documentation:

None.

#### Reviewer Suggestions: 

Anyone.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Rob Trangucci

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
